### ISSUE TITLE
18TN auto route uses the wrong trains during Civil War

### ISSUE THREAD

**Issue by ChrisTheCat** [tobymao/18xx#6597]
https://18xx.games/game/63972?action=187

The GMO auto route would be better if it used a 2 and a 3, leaving a 2 unused for the Civil War. Instead, it uses both 2 trains and applies the Civil War penalty to the 3 train.

**Comment by mamitsch** [comment]
Same issue happens on 1889 - The IR had two cities connected to the border tile and three trains (2 2 3), but only one token. Autoroute solves this with only the 2 trains, instead of using one 2 and one 3 train.

**Comment by perwestling** [comment]
> Same issue happens on 1889 - The IR had two cities connected to the border tile and three trains (2 2 3), but only one token. Autoroute solves this with only the 2 trains, instead of using one 2 and one 3 train.

Would the 2 train + 3 train score more than a 2 train + 2 train?  In the case of 18TN above the 3 train and 2 train would just make two stops so it did not matter which train was ignored.

**Comment by mamitsch** [comment]
Check hot seat

Id: hs_hhffojge_1641487070



With the 3 you can earn 20+30 for the 2 and 30 + 20 + 20 for the 3, together 120

Auto resolves wrong - only 90.

Best,

Marcus

**Comment by perwestling** [comment]
@mamitsch Hot seat games are local to your browser. So you need to export it. I believe there are instructions for that in the issue template.

I know there is a discussion about auto routing in the 18xxgamesdev chat right now. I believe the 1889 thing might be a different issue (a general one) while the 18TN is connected to the Civil War effect. Now I am redesigning 18TN to make it work better with autorouting, but I think the 1889 thing might be a good reason to add a new issue.

**Comment by mamitsch** [comment]
Hi there,

Here is the file of the game



Sent from my iPhone

> On 6. Jan 2022, at 17:58, Per Westling ***@***.***> wrote:
> 
> ﻿
> @mamitsch Hot seat games are local to your browser. So you need to export it. I believe there are instructions for that in the issue template.
> 
> I know there is a discussion about auto routing in the 18xxgamesdev chat right now. I believe the 1889 thing might be a different issue (a general one) while the 18TN is connected to the Civil War effect. Now I am redesigning 18TN to make it work better with autorouting, but I think the 1889 thing might be a good reason to add a new issue.
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> Triage notifications on the go with GitHub Mobile for iOS or Android. 
> You are receiving this because you were mentioned.

{
  "status": "active",
  "actions": [
    {
      "type": "bid",
      "entity": "Player 1",
      "entity_type": "player",
      "id": 1,
      "created_at": 1641487075,
      "company": "TR",
      "price": 20
    },
    {
      "type": "bid",
      "entity": "Player 2",
      "entity_type": "player",
      "id": 2,
      "created_at": 1641487076,
      "company": "MF",
      "price": 30
    },
    {
      "type": "bid",
      "entity": "Player 3",
      "entity_type": "player",
      "id": 3,
      "created_at": 1641487091,
      "company": "ER",
      "price": 40
    },
    {
      "type": "pass",
      "entity": "Player 1",
      "entity_type": "player",
      "id": 4,
      "created_at": 1641487093
    },
    {
      "type": "pass",
      "entity": "Player 2",
      "entity_type": "player",
      "id": 5,
      "created_at": 1641487094
    },
    {
      "type": "bid",
      "entity": "Player 3",
      "entity_type": "player",
      "id": 6,
      "created_at": 1641487096,
      "company": "SMR",
      "price": 50
    },
    {
      "type": "pass",
      "entity": "Player 1",
      "entity_type": "player",
      "id": 7,
      "created_at": 1641487096
    },
    {
      "type": "pass",
      "entity": "Player 2",
      "entity_type": "player",
      "id": 8,
      "created_at": 1641487097
    },
    {
      "type": "bid",
      "entity": "Player 3",
      "entity_type": "player",
      "id": 9,
      "created_at": 1641487098,
      "company": "DR",
      "price": 60
    },
    {
      "type": "pass",
      "entity": "Player 1",
      "entity_type": "player",
      "id": 10,
      "created_at": 1641487099
    },
    {
      "type": "pass",
      "entity": "Player 2",
      "entity_type": "player",
      "id": 11,
      "created_at": 1641487099
    },
    {
      "type": "bid",
      "entity": "Player 3",
      "entity_type": "player",
      "id": 12,
      "created_at": 1641487101,
      "company": "SIR",
      "price": 80
    },
    {
      "type": "par",
      "entity": "Player 1",
      "entity_type": "player",
      "id": 13,
      "created_at": 1641487107,
      "corporation": "IR",
      "share_price": "80,2,3"
    },
    {
      "type": "par",
      "entity": "Player 2",
      "entity_type": "player",
      "id": 14,
      "created_at": 1641487113,
      "corporation": "AR",
      "share_price": "75,3,3"
    },
    {
      "type": "buy_shares",
      "entity": "Player 3",
      "entity_type": "player",
      "id": 15,
      "created_at": 1641487120,
      "shares": [
        "IR_1"
      ],
      "percent": 10
    },
    {
      "type": "buy_shares",
      "entity": "Player 1",
      "entity_type": "player",
      "id": 16,
      "created_at": 1641487124,
      "shares": [
        "IR_2"
      ],
      "percent": 10
    },
    {
      "type": "buy_shares",
      "entity": "Player 2",
      "entity_type": "player",
      "id": 17,
      "created_at": 1641487134,
      "shares": [
        "AR_1"
      ],
      "percent": 10
    },
    {
      "type": "buy_shares",
      "entity": "Player 3",
      "entity_type": "player",
      "id": 18,
      "created_at": 1641487140,
      "shares": [
        "IR_3"
      ],
      "percent": 10
    },
    {
      "type": "buy_shares",
      "entity": "Player 1",
      "entity_type": "player",
      "id": 19,
      "created_at": 1641487143,
      "shares": [
        "IR_4"
      ],
      "percent": 10
    },
    {
      "type": "buy_shares",
      "entity": "Player 2",
      "entity_type": "player",
      "id": 20,
      "created_at": 1641487146,
      "shares": [
        "AR_2"
      ],
      "percent": 10
    },
    {
      "type": "buy_shares",
      "entity": "Player 1",
      "entity_type": "player",
      "id": 21,
      "created_at": 1641487148,
      "shares": [
        "IR_5"
      ],
      "percent": 10
    },
    {
      "type": "buy_shares",
      "entity": "Player 2",
      "entity_type": "player",
      "id": 22,
      "created_at": 1641487150,
      "shares": [
        "AR_3"
      ],
      "percent": 10
    },
    {
      "type": "lay_tile",
      "entity": "IR",
      "entity_type": "corporation",
      "id": 23,
      "created_at": 1641487162,
      "hex": "E2",
      "tile": "5-0",
      "rotation": 4
    },
    {
      "type": "buy_train",
      "entity": "IR",
      "entity_type": "corporation",
      "id": 24,
      "created_at": 1641487165,
      "train": "2-0",
      "price": 80,
      "variant": "2"
    },
    {
      "type": "buy_train",
      "entity": "IR",
      "entity_type": "corporation",
      "id": 25,
      "created_at": 1641487165,
      "train": "2-1",
      "price": 80,
      "variant": "2"
    },
    {
      "type": "buy_train",
      "entity": "IR",
      "entity_type": "corporation",
      "id": 26,
      "created_at": 1641487174,
      "train": "2-2",
      "price": 80,
      "variant": "2"
    },
    {
      "type": "pass",
      "entity": "IR",
      "entity_type": "corporation",
      "id": 27,
      "created_at": 1641487177
    },
    {
      "type": "lay_tile",
      "entity": "AR",
      "entity_type": "corporation",
      "id": 28,
      "created_at": 1641487188,
      "hex": "K8",
      "tile": "6-0",
      "rotation": 2
    },
    {
      "type": "buy_train",
      "entity": "AR",
      "entity_type": "corporation",
      "id": 29,
      "created_at": 1641487190,
      "train": "2-3",
      "price": 80,
      "variant": "2"
    },
    {
      "type": "buy_train",
      "entity": "AR",
      "entity_type": "corporation",
      "id": 30,
      "created_at": 1641487190,
      "train": "2-4",
      "price": 80,
      "variant": "2"
    },
    {
      "type": "buy_train",
      "entity": "AR",
      "entity_type": "corporation",
      "id": 31,
      "created_at": 1641487192,
      "train": "2-5",
      "price": 80,
      "variant": "2"
    },
    {
      "type": "buy_train",
      "entity": "AR",
      "entity_type": "corporation",
      "id": 32,
      "created_at": 1641487202,
      "train": "3-0",
      "price": 180,
      "variant": "3"
    },
    {
      "type": "pass",
      "entity": "AR",
      "entity_type": "corporation",
      "id": 33,
      "created_at": 1641487207
    },
    {
      "type": "buy_shares",
      "entity": "DR",
      "entity_type": "company",
      "id": 34,
      "created_at": 1641487219,
      "shares": [
        "IR_6"
      ],
      "percent": 10
    },
    {
      "type": "buy_shares",
      "entity": "Player 3",
      "entity_type": "player",
      "id": 35,
      "created_at": 1641487221,
      "shares": [
        "IR_7"
      ],
      "percent": 10
    },
    {
      "type": "pass",
      "entity": "Player 3",
      "entity_type": "player",
      "id": 36,
      "created_at": 1641487229
    },
    {
      "type": "pass",
      "entity": "Player 1",
      "entity_type": "player",
      "id": 37,
      "created_at": 1641487230
    },
    {
      "type": "pass",
      "entity": "Player 2",
      "entity_type": "player",
      "id": 38,
      "created_at": 1641487234
    },
    {
      "type": "pass",
      "entity": "Player 3",
      "entity_type": "player",
      "id": 39,
      "created_at": 1641487236
    },
    {
      "type": "lay_tile",
      "entity": "IR",
      "entity_type": "corporation",
      "id": 40,
      "created_at": 1641487243,
      "hex": "F3",
      "tile": "5-1",
      "rotation": 2
    },
    {
      "type": "pass",
      "entity": "IR",
      "entity_type": "corporation",
      "id": 41,
      "created_at": 1641487249
    },
    {
      "type": "run_routes",
      "entity": "IR",
      "entity_type": "corporation",
      "id": 42,
      "created_at": 1641487260,
      "routes": [
        {
          "train": "2-0",
          "connections": [
            [
              "E2",
              "F3"
            ]
          ],
          "hexes": [
            "E2",
            "F3"
          ],
          "revenue": 40,
          "revenue_str": "E2-F3"
        },
        {
          "train": "2-1",
          "connections": [
            [
              "E2",
              "F1"
            ]
          ],
          "hexes": [
            "E2",
            "F1"
          ],
          "revenue": 50,
          "revenue_str": "E2-F1"
        }
      ]
    },
    {
      "type": "dividend",
      "entity": "IR",
      "entity_type": "corporation",
      "id": 43,
      "created_at": 1641487263,
      "kind": "payout"
    },
    {
      "type": "buy_train",
      "entity": "IR",
      "entity_type": "corporation",
      "id": 44,
      "created_at": 1641487264,
      "train": "3-1",
      "price": 180,
      "variant": "3"
    },
    {
      "type": "pass",
      "entity": "IR",
      "entity_type": "corporation",
      "id": 45,
      "created_at": 1641487274
    },
    {
      "type": "lay_tile",
      "entity": "AR",
      "entity_type": "corporation",
      "id": 46,
      "created_at": 1641487298,
      "hex": "K8",
      "tile": "12-0",
      "rotation": 2
    },
    {
      "type": "run_routes",
      "entity": "AR",
      "entity_type": "corporation",
      "id": 47,
      "created_at": 1641487308,
      "routes": [
        {
          "train": "2-3",
          "connections": [
            [
              "K8",
              "L7"
            ]
          ],
          "hexes": [
            "K8",
            "L7"
          ],
          "revenue": 50,
          "revenue_str": "K8-L7"
        }
      ]
    },
    {
      "type": "dividend",
      "entity": "AR",
      "entity_type": "corporation",
      "id": 48,
      "created_at": 1641487311,
      "kind": "payout"
    },
    {
      "type": "pass",
      "entity": "AR",
      "entity_type": "corporation",
      "id": 49,
      "created_at": 1641487312
    },
    {
      "type": "pass",
      "entity": "IR",
      "entity_type": "corporation",
      "id": 50,
      "created_at": 1641487325
    },
    {
      "type": "pass",
      "entity": "IR",
      "entity_type": "corporation",
      "id": 51,
      "created_at": 1641487327
    }
  ],
  "id": "hs_hhffojge_1641487070",
  "players": [
    {
      "name": "Player 1"
    },
    {
      "name": "Player 2"
    },
    {
      "name": "Player 3"
    }
  ],
  "title": "1889",
  "description": "",
  "max_players": "3",
  "settings": {
    "optional_rules": []
  },
  "mode": "hotseat",
  "user": {
    "id": 0,
    "name": "You"
  },
  "created_at": "2022-01-06",
  "loaded": true,
  "result": {},
  "turn": 2,
  "round": "Operating Round",
  "acting": [
    "Player 1"
  ],
  "updated_at": 1641487327
}

**Comment by jcoveney** [comment]
I ran into this same issue in 1830, so I don't think it has to do with the Civil War.
Here is a gist of a hotseat game: https://gist.github.com/jcoveney/09595bf47a18bc271d8caf9388bad0c8
Same idea...it should use the 3, but instead uses all 2Ts.

**Comment by perwestling** [comment]
Have now created a new issue (#6774) where I used the 1889 and 1830 games as reference.

The 18TN / Civil War issue I still needs some reimplementation specific to 18TN, as the current implementation does not work well with Civil War and auto routing.

**Comment by tobymao** [comment]
just use this thread to track

**Comment by ChrisTheCat** [comment]
Another example: https://18xx.games/game/67804?action=303

<img width="344" alt="Screen Shot 2022-01-17 at 08 50 14" src="https://user-images.githubusercontent.com/8732435/149809578-ed52f71a-73c6-42f8-84c3-7402cd3eb23b.png">

<img width="324" alt="Screen Shot 2022-01-17 at 08 50 06" src="https://user-images.githubusercontent.com/8732435/149809583-87d2a395-eee8-4b58-8a9a-23e5a5794747.png">

**Comment by JoeB989** [comment]
I retested with the current state of auto router, and I believe that the 1830, 1861, and 1889 games now route properly.  The 18TN game still doesn't favor the 3 over the 2.  (This is due to the order they were purchased, and only having 2-routes)

@perwestling , I have three ideas for an 18TN solution:
1. Include the civil war penalty in Game::routes_revenue (if that's possible).  Then the auto-router should always select the best value routes
2. I had played with auto-router sorting trains by length, but ended up removing it.  This is a general solution but might solve your specific 18TN issue.
3. I recently added the ability for Game::check_other to reject "valid but not valid" combos.  I used this for 1870 destination run constraints, but you might use it like this:
   - 	if some trains are unused, reject combos where they're not the shortest trains

Also, is the computed value supposed to be the same whether the 3 or 2 doesn't run?  If the value was less with the 2, then auto-router should always select the 3.
